parcels/interaction/interactionkernelsoa.py

```diff
@@ -0,0 +1,179 @@
+from parcels.field import VectorField
+from parcels.interaction.baseinteractionkernel import BaseInteractionKernel
+import parcels.rng as ParcelsRandom  # noqa
+from parcels.tools.statuscodes import StateCode, OperationCode, ErrorCode
+from parcels.tools.statuscodes import recovery_map as recovery_base_map
+from parcels.tools.loggers import logger
+
+
+__all__ = ['InteractionKernelSOA']
+
+
+class InteractionKernelSOA(BaseInteractionKernel):
+    """InteractionKernel object that encapsulates auto-generated code.
+    """
+
+    def __init__(self, fieldset, ptype, pyfunc=None, funcname=None,
+                 funccode=None, py_ast=None, funcvars=None, c_include="",
+                 delete_cfiles=True):
+        super().__init__(fieldset=fieldset, ptype=ptype, pyfunc=pyfunc,
+                         funcname=funcname, funccode=funccode, py_ast=py_ast,
+                         funcvars=funcvars, c_include=c_include,
+                         delete_cfiles=delete_cfiles)
+
+        for func in self._pyfunc:
+            self.check_fieldsets_in_kernels(func)
+
+        numkernelargs = self.check_kernel_signature_on_version()
+
+        assert numkernelargs[0] == 5 and \
+            numkernelargs.count(numkernelargs[0]) == len(numkernelargs), \
+            'Interactionkernels take exactly 5 arguments: particle, fieldset, time, neighbors, mutator'
+
+        # At this time, JIT mode is not supported for InteractionKernels,
+        # so there is no need for any further "processing" of pyfunc's.
+
+    def execute_jit(self, pset, endtime, dt):
+        raise NotImplementedError
+
+    def __del__(self):
+        # Clean-up the in-memory dynamic linked libraries.
+        # This is not really necessary, as these programs are not that large, but with the new random
+        # naming scheme which is required on Windows OS'es to deal with updates to a Parcels' kernel.)
+        super().__del__()
+
+    def __add__(self, kernel):
+        if not isinstance(kernel, InteractionKernelSOA):
+            kernel = InteractionKernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return self.merge(kernel, InteractionKernelSOA)
+
+    def __radd__(self, kernel):
+        if not isinstance(kernel, InteractionKernelSOA):
+            kernel = InteractionKernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return kernel.merge(self, InteractionKernelSOA)
+
+    def execute_python(self, pset, endtime, dt):
+        """Performs the core update loop via Python"""
+        if self.fieldset is not None:
+            for f in self.fieldset.get_fields():
+                if type(f) in [VectorField, NestedField, SummedField]:
+                    continue
+                f.data = np.array(f.data)
+
+        for pyfunc in self._pyfunc:
+            pset.compute_neighbor_tree(endtime, dt)
+            active_idx = pset._active_particle_idx
+
+            mutator = defaultdict(lambda: [])
+            for particle_idx in active_idx:
```

Comment:
Here again, it would be extremely useful what `_active_idx` includes or represents.

Reply:
See above

Reply:
can you please point me to file and function that refers to 'above' ?